### ISSUE TITLE
Encouraging Given/When/Then

### ISSUE THREAD

**Issue by ciaranmcnulty** [phpspec/phpspec#628]
An optional mechanism to structure tests into Given/When/Then syntax would be helpful for new learners; it would not ever be required.

I'm unsure how this can be achieved nicely in a PHP syntax so input is welcome.

Something like this, but more concise:

```php
function it_should_do_something() {
    return [
      self::GIVEN = function () { ... }
      self::WHEN = function () { ... }
      self::THEN = function () { ... }
    ];
}
```

We could enforce which actions (stubbing, matchers, method invocations) are allowed to be used in each section as a way to get people used to the Setup/Exercise/Verify or Assemble/Act/Assert structure of good tests.

**Comment by henrikbjorn** [comment]
Well, doesnt this already exist through behat? and isnt behat a better platform for these kind of tests?

**Comment by ciaranmcnulty** [comment]
We don't have to call them Given/When/Then if it's going to make people think of acceptance tests, we can use another naming.

But the idea of structuring a unit testing into 3 (or 4, if you include teardown/annihilate) steps long predates Behat and indeed BDD (e.g. https://weblogs.java.net/blog/wwake/archive/2003/12/tools_especiall.html).

When learning to write good Unit Tests (which are different to specs but close enough) it's a good structure to stick to. As your understanding of the subject matures it's fine to drop it.

What I'm after in a more generic sense is 'a way to help newcomers to structure their tests correctly'

**Comment by jakzal** [comment]
There's a name for this that predates Given-When-Then - it's [Arrange - Act - Assert](http://c2.com/cgi/wiki?ArrangeActAssert).

Not sure if wrapping these steps in anonymous functions would help beginners. It increases number of indentation and makes specs less readable. 

Alternatively, we could generate the first spec with appropriate comments

```php5
function it_should_do_something()
{
    // Arrange

    // Act

    // Assert
}
```

**Comment by ciaranmcnulty** [comment]
@jakzal I know about AAA, I linked to an article about it above ;-)

Comments are a nice idea! Would it annoy experienced users (more than the existing TODO)?

The anonymous functions are not a serious suggestion; it may be something cleaner isn't possible in PHP :(

Having it marked into blocks would enable us to enforce what can be done inside each phase, again I think that would help new users.

**Comment by jon-acker** [comment]
I tend to agree with @jakzal on this, if the idea is basically to help new learners identify the different parts of the test/spec (AAA).

**Comment by ciaranmcnulty** [comment]
Another option for a strict mode would be to enforce the order in which certain actions can be taken.

e.g. throw an error if a method is mocked after a matcher has been used.

**Comment by MarcelloDuarte** [comment]
Very often PhpSpec will do Arrange for you and you can Act and Assert in one line anyway, which I think is an elegant aspect of the tool.

**Comment by ciaranmcnulty** [comment]
@MarcelloDuarte for descriptiveness one-liners are great, my main motivation was to try and avoid some of the common antipatterns I've seen recently (e.g. act/assert/act/assert structures)

**Comment by ciaranmcnulty** [comment]
So for instance, maybe using a matcher should prevent any more stubbing or mocking?

**Comment by jakzal** [comment]
@ciaranmcnulty :+1: that would be nice.  Would there be any issues with spies?

We could also document the common anti patterns.

**Comment by ciaranmcnulty** [comment]
We'd have to explicitly allow spies I guess.

**Comment by ciaranmcnulty** [comment]
@jakzal Having thought about it, spies would go into the Assert categories with Matchers.

So as soon as you call a matcher or invoke a Spy, any Mocking or Stubbing is forbidden.

**Comment by dmecke** [comment]
I would call myself a learner about how to structure unit tests in a good way, so as this issue is meant for people like me I would like to tell my view on it.

First, I really wasnt aware that I should mock all dependencies, not just the ones I absolutely need to get the tests running (I think I mainly actually wrote integration tests). phpspec made this immediately clear as I have no other chance.
So disallow mocking after an assertion seems very close to this for me. Though a different approach to this could be to instead mark the test yellow with a warning that this is not the intended way to do it.

**Comment by everzet** [comment]
Here's my syntax proposal:

```
function it_should_do_something() {
    // given
    ...
    // when
    ...
    // then
    ...
}
```

**Comment by ciaranmcnulty** [comment]
@dmecke You shouldn't mock all dependencies ;-)

**Comment by dmecke** [comment]
@ciaranmcnulty This seems to become offtopic, but aren't it integration tests then? I mean, my SUT then depends on behaviour of its collaborators!?

**Comment by shanethehat** [comment]
It would be possible, no matter whether comments or something more complex is used, to have a config option that disables and pre-written body to your spec. That would appease experienced users who don't want the hassle of deleting.

**Comment by everzet** [comment]
@dmecke your SUT **always** depends on behaviour of its collaborators. We don't use mocks to pretend that's not the case, we use mocks to expose, inspect and improve these dependencies. Not all dependencies need exposure, inspection and improvement.

**Comment by bestform** [comment]
@everzet @ciaranmcnulty I think, @dmecke is right. Not mocking all dependencies does make the spec a secret integration test. The results of that can be pretty unpleasant. Not only is it not the purpose of the spec of class A to ensure the correct behaviour of a dependency B. Even worse: if class B breaks, both specs (for class A and class B) break, making it more difficult to identify the problem immediately.

If class B breaks, only the spec for class B should break. Class A has nothing to do with this error and therefore should still work in the test environment.

(Sorry for driving this slightly off topic discussion further, but in my opinion this is an important point)

**Comment by fesor** [comment]
@bestform 

> Not mocking all dependencies does make the spec a secret integration test.

Mocking all dependencies without any exception makes your test highly coupled with current implementation. 

> Not only is it not the purpose of the spec of class A to ensure the correct behaviour of a dependency B

This dependency should be hidden, we are testing only object A behaviour, we don't know anything about B. It could use whatever we want to have behaviour that we need to get.

> if class B breaks, both specs (for class A and class B) break, making it more difficult to identify the problem immediately.

You should know what went wrong not by "what failed" but more from "what changed". If I change a few lines of code and my tests are failing, I pretty much know what went wrong. It's just a matter of feedback loop length.

p.s. Sorry for offtopic.

**Comment by jakzal** [comment]
Sorry for off topic again, but I need to clarify this ;)

> You should alway mock third-party dependencies

You should (almost) never mock your third party dependencies. For code that uses third party objects integration tests should be written. Otherwise you'd be making assumptions without really making sure they're true.

**Comment by fesor** [comment]
@jakzal well... you right)

But then you shouldn't write such tests with phpspec since it really only good in unit testing.
